scipy/sparse/_csparsetools.pyx.in

```diff
@@ -461,3 +461,30 @@ cdef _fill_dtype_map2(map):
 
 _fill_dtype_map(_LIL_FANCY_GET_DISPATCH, np.typecodes['Integer'])
 _fill_dtype_map2(_LIL_FANCY_SET_DISPATCH)
+
+
+def lil_matvec(int M, int N,
+               object[:] rows,
+               object[:] data,
+               cnp.ndarray vec,
+               cnp.ndarray result):
+    """
+    Compute y = A * x for LIL matrix A and dense vectors x and y.
+
+    Parameters
+    ----------
+    M, N, rows, data
+        The data from the LIL matrix A.
+    vec
+        The vector x as array.
+    res
+        The vector y as array.
+        Must be preallocated to shape `N`!
+
+    """
+
+    for i in range(M):
+        result_i = 0
+        for (j, A_ij) in zip(rows[i], data[i]):
```

Comment:
I'm not sure if Cython handles this well or not. It might be moot if object array indexing is the bottleneck, though.

Reply:
So should I change it to something like this?
```
for k in range(len(rows)):
    result_i += data[k] * vec[rows[k]]

Reply:
A note from the peanut gallery: Cython does not handle += for complex-valued arguments.

Reply:
Nice to know. But then I am surprised that the added `test_mul_vector` test did not produce any errors.

So I will change it to :
```
result_i = result_i + data[k] * vec[rows[k]]
```